### PR TITLE
Verify doc examples

### PR DESCRIPTION
**Pull request by grahamc**
###### Motivation for this change

Test the function examples actually work.

Testing steps:

1. enter doc/
2. nix-shell
3. make doc-tests

It outputs something like:

```
$ make doc-tests 
cd ./functions/library; python3 ./test.py
function-library-lib.attrset.attrByPath-example-value-exists: pass
function-library-lib.attrset.attrByPath-example-default-value: pass
function-library-lib.attrsets.recursiveUpdateUntil-example: pass
function-library-lib.attrsets.recursiveUpdate-example: pass
function-library-lib.attrsets.matchAttrs-example-matching-pattern-deep: pass
5 passed, 0 failed

```

For each example without a testable program, it prints:

```
Skipping example function-library-lib.attrsets.zipAttrs-example: no <programlisting role="input">
```

If a test fails, it outputs instructions on how to test only the failing examples:

```
function-library-lib.attrsets.recursiveUpdateUntil-example: pass
function-library-lib.attrsets.recursiveUpdate-example: pass
--------  Doc test failure  --------
ID: function-library-lib.attrsets.matchAttrs-example-matching-pattern-shallow
Executing:


      {
        actual = (let lib = import /home/grahamc/projects/nixpkgs/lib; in (
          
lib.attrsets.matchAttrs
  { cpu = {}; }
  { cpu = { bits = 64; }; }

        ));
        expect = (
          
  false

        );
      }
    

Expecting:

  false

Found:
True
----
--------  Doc test failure  --------
ID: function-library-lib.attrsets.matchAttrs-example-matching-pattern-deep
Executing:


      {
        actual = (let lib = import /home/grahamc/projects/nixpkgs/lib; in (
          
lib.attrsets.matchAttrs
  { host = { cpu = {}; }; }
  { host = { cpu = { bits = 64; }; }; }

        ));
        expect = (
          false
        );
      }
    

Expecting:
false
Found:
True
----
4 passed, 2 failed
Retry failing examples with:
./functions/library/test.py --lib-dir /home/grahamc/projects/nixpkgs/lib function-library-lib.attrsets.matchAttrs-example-matching-pattern-shallow function-library-lib.attrsets.matchAttrs-example-matching-pattern-deep
```

These tests will block the Nixpkgs manual from building as well.

The XML isn't super nice to write, but I think isn't too onerous?

Anyway, rendered it looks like this: 

image

I also verified that python3 was already in the build closure.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: doc/functions/library/test.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        expect = example.find('.//docbook:programlisting[@role="output"]', ns)
+        if expect is None:
+            debug('Skipping example {}: no <programlisting role="output">'.format(xml_id))
+            continue
+
+        yield { "id": xml_id, "code": code.text, "expect": expect.text }
+
+
+parser = argparse.ArgumentParser(description='Execute doc tests for the Nixpkgs manual..')
```
Comment (FRidh) [inline doc/functions/library/test.py]: put the following code in a function, e.g. `main()` and add `if __name __ = "__main__": main()

File: doc/functions/library/test.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                    help='Source of the library to test against (default: {})'.format(default_lib_dir))
+parser.add_argument('ids', metavar='id', type=str, nargs='*',
+                    help='an example ID to test')
+
+args = parser.parse_args()
+
+doc_root = os.path.dirname(os.path.realpath(__file__))
+all_doc_files = glob.glob("{}/*.xml".format(doc_root))
+
+all_docs = [ET.parse(filename) for filename in all_doc_files]
```
Comment (FRidh) [inline doc/functions/library/test.py]: generator expression

File: doc/functions/library/test.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+all_examples = itertools.chain(*[find_examples(doc) for doc in all_docs])
+
+if len(args.ids) > 1:
```
Comment (FRidh) [inline doc/functions/library/test.py]: if args.ids:

File: doc/default.nix [hunk 4]
```diff
@@ -6,7 +6,8 @@ in
 pkgs.stdenv.mkDerivation {
   name = "nixpkgs-manual";
 
-  buildInputs = with pkgs; [ pandoc libxml2 libxslt zip jing  xmlformat ];
+  buildInputs = with pkgs; [ pandoc libxml2 libxslt zip jing xmlformat
+    python36 nix ];
```
Comment (FRidh) [inline doc/default.nix]: python3 is sufficient. Note that soon I intend to make python37 the default, and this code will still work fine then.

File: doc/functions/library/test.py (hunk 2 above)
Comment (graham-at-target) [reply doc/functions/library/test.py]: Hmm I'm not sure what to do with this feedback?

File: doc/functions/library/test.py (hunk 3 above)
Comment (shlevy) [reply doc/functions/library/test.py]: I don't think so, this is >1 not >0

Comment (shlevy) [reply doc/functions/library/test.py]: ... But should it be? seems like we drop a singleton list here? OK agreed with @FRidh then

File: doc/functions/library/test.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        all_examples
+    )
+else:
+    filtered_examples = all_examples
+
+passed = 0
+failed = []
+
+for example in filtered_examples:
+    full_expression = """
```
Comment (shlevy) [inline doc/functions/library/test.py]: Maybe want [f strings](https://www.python.org/dev/peps/pep-0498/)?

**Review by shlevy — APPROVED** [review]
Nice!

File: doc/functions/library/test.py (hunk 3 above)
Comment (graham-at-target) [reply doc/functions/library/test.py]: +1 this is a relic of when I didn't have proper argument parsing 😁 good catch.

**Comment by mmahut** [comment]
Are there any updates on this pull request, please?

**Comment by asymmetric** [comment]
@grahamc has this been abandoned?

**Pull request closed without merging**